xdp-dump/xdpdump.c

```diff
@@ -416,27 +416,119 @@ static bool capture_on_legacy_interface(struct dumpopt *cfg)
 	return rc;
 }
 
+static size_t find_func_matches(const struct btf *btf,
+				const char *func_name,
+				const char **found_name,
+				bool print)
+{
+	const struct btf_type *t, *match;
+	size_t len, matches = 0;
+	const char *name;
+	int nr_types, i;
+
+	if (!btf) {
+		pr_debug("No BTF found for program\n");
+		return 0;
+	}
+
+	len = strlen(func_name);
+
+	nr_types = btf__get_nr_types(btf);
+	for (i = 1; i <= nr_types; i++) {
+		t = btf__type_by_id(btf, i);
+		if (!btf_is_func(t))
+			continue;
+
+		name = btf__name_by_offset(btf, t->name_off);
+		if (!strncmp(name, func_name, len)) {
+			pr_debug("Found func %s matching %s\n",
+				 name, func_name);
+
+			if (print)
+				pr_warn("  %s\n", name);
+
+			if (strlen(name) == len) {
```

Comment:
This should be removed 

```
			/* if (strlen(name) == len) { */
			/* 	*found_name = name; */
			/* 	return 1; /\* exact match *\/ */
			/* } */
```

As it's failing the test case where the exact name is not the function loaded.

Reply:
Not sure I understand your comment; what is failing?

Reply:
If I have three functions in my BTF (file) and one is an exact match but not the one actually being used it assumes this is the one needed. Which might not be the case, its better to fail and let the user pick the correct one and not assume the exact match is the correct one.

Reply:
But how is a user supposed to specify the right candidate if an exact match gets rejected when it's also a prefix of another function?

I.e., say there's `funcname1()` and `funcname1_more()` and the user **wants** to use `funcname1()`, if we don't have this 'exact match' short-circuiting that won't be possible, unless I'm missing something?

Reply:
You are right, that test case I did before I made the change :(

The exact match should only be valid for the -p option, and probably also when the length is less than 15 chars as than no collision can happen. Think I had that in the original code.

Reply:
Yeah, you did. The reason I removed it was that the name is now expanded by libxdp on load, so the truncated name will only survive if that initial expansion is ambiguous. Hmm, I guess we could have if (-p set || len != 15) to automatically match both longer and shorter names in one go, then?